File: go/leetcode/144. binary tree preorder traversal_iterative.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
    arr := []int {}
  
    if root == nil {
        return arr;
    }
    
    stack := []*TreeNode{}
    stack = append(stack, root)
    
    for len(stack) > 0 {
        top := stack[ len(stack) - 1 ]
        arr = append(arr, top.Val)
        stack = stack[ :len(stack) - 1 ]
        
        fmt.Println(top.Val)
        
        if top.Right != nil {
            stack = append(stack, top.Right)
        } 
        if top.Left != nil {
            stack = append(stack, top.Left)
        }
    }
    return arr;
}

File: c++/leetcode/144. binary tree preorder traversal_iterative.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  vector<int> preorderTraversal(TreeNode *root)
  {
    vector<int> array;

    if (root == nullptr)
      return array;

    stack<TreeNode *> stk;
    stk.push(root);

    while (!stk.empty())
    {
      TreeNode *removed = stk.top();
      stk.pop();
      array.push_back(removed->val);

      if (removed->right)
        stk.push(removed->right);
      if (removed->left)
        stk.push(removed->left);
    }

    return array;
  }
};
// /**
//  * Definition for a binary tree node.
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
//  * };
//  */
// class Solution
// {
// public:
//   vector<int> preorderTraversal(TreeNode *root)
//   {
//     vector<int> array;
//     if (root == nullptr)
//       return array;

//     stack<TreeNode *> stk;
//     stk.push(root);

//     while (!stk.empty())
//     {
//       // pop
//       TreeNode *removed = stk.top();
//       stk.pop();
//       array.push_back(removed->val);

//       // push if right
//       if (removed->right)
//       {
//         stk.push(removed->right);
//         preorderTraversal(removed->right);
//       }

//       // push if left
//       if (removed->left)
//       {
//         stk.push(removed->left);
//         preorderTraversal(removed->left);
//       }
//     }

//     return array;
//   }
// };


